io-classes/src/Control/Monad/Class/MonadMVar.hs

```diff
@@ -0,0 +1,379 @@
+                   )
+                => MVarDefault m a
+                -> m a
+takeMVarDefault (MVar tv) = mask_ $ do
+    res <- atomically $ do
+      s <- readTVar tv
+      case s of
+        -- if it's empty we add ourselves to the blocked queue
+        MVarEmpty blockedq -> do
+          wakevar <- newTVar Nothing
+          writeTVar tv (MVarEmpty (Deque.snoc wakevar blockedq))
+          return (Left wakevar)
+
+        -- if it's full we grab the value, and also complete the action of the
+        -- next thread blocked in putMVar, by setting the new MVar value and
+        -- unblocking them.
+        MVarFull x blockedq ->
+          case Deque.uncons blockedq of
+            Nothing ->
+              return (Right x)
+
+            Just ((x', wakevar), blockedq') -> do
+              writeTVar wakevar True
+              writeTVar tv (MVarFull x' blockedq')
+              return (Right x)
+
+    case res of
+      -- we have to block on our own wakevar until we can complete the read
+      Left wakevar ->
+        atomically (readTVar wakevar >>= maybe retry return)
+        `catch` \e@SomeAsyncException {} -> do
+          atomically $ do
+            s <- readTVar tv
+            case s of
+              MVarEmpty blockedq -> do
+                -- async exception was thrown while were were blocked on
+                -- wakevar; we need to remove it from 'blockedq', otherwise we
+                -- will have a space leak.
+                let blockedq' = Deque.filter (/= wakevar) blockedq
+                writeTVar tv (MVarEmpty blockedq')
+              -- the exception was thrown while we were blocked on 'wakevar', so
+              -- the 'MVar' must not be full.
+              MVarFull {} -> error "takeMVarDefault: invariant violation"
+          throwIO e
+
+      -- we managed to do the take synchronously
+      Right x -> return x
+
+
+-- | 'readMVarDefault' when the 'MVar' is empty, guarantees to receive next
+-- 'putMVar' value.  It will also not block if the 'MVar' is full, even if there
+-- are other threads attempting to 'putMVar'.
+--
+readMVarDefault :: MonadSTM m
+                => MVarDefault m a
+                -> m a
+readMVarDefault (MVar tv) = do
+    atomically $ do
+      s <- readTVar tv
+      case s of
+        -- if it's empty block
+        MVarEmpty _  -> retry
+
+        -- if it's full return the value
+        MVarFull x _ -> return x
+
+
+tryTakeMVarDefault :: MonadSTM m
+                   => MVarDefault m a
+                   -> m (Maybe a)
+tryTakeMVarDefault (MVar tv) = do
+    atomically $ do
+      s <- readTVar tv
+      case s of
+        MVarEmpty _ -> return Nothing
+        MVarFull x blockedq ->
+          case Deque.uncons blockedq of
+            Nothing -> return (Just x)
+            Just ((x', wakevar), blockedq') -> do
+              writeTVar wakevar True
+              writeTVar tv (MVarFull x' blockedq')
+              return (Just x)
```

Comment:
Might be worth to generalize this to something like `takeSTM` and reuse in `takeMVarDefault` as well

Reply:
this one does not need to be masked, but yeah maybe can share some code.

Reply:
I am not sure if it's worth it. The first atomic block of `takeMVarDefault` and `tryTakeMVar` differs in the `MVarEmpty` case, so we will still need to match on the result of `takeSTM`.  If we do that will also provide an indirection which is not that convenient when reading the code.

Reply:
Agree! 😄